lib/state_machine/types.hpp

```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include "message.hpp"
+#include "state.hpp"
+
+#include <cstdint>
+
+#include "utils/hash.hpp"
+
+namespace hyped::state_machine {
+
+struct TransitionKey {
```

Comment:
This could just live inside `state_machine::StateMachine`? It's not going to be used anywhere else and it's just for the internals.

I'm not sure whether that's possible, but I believe you can make it private. Might be an issue with the `std::` template, though...